msprime/demography.py

```diff
@@ -3459,18 +3459,36 @@ def _proportions_to_sequential(P):
     return C
 
 
-def _matrix_exponential(A):
+def _matrix_exponential(A, n=5):
     """
     Returns the matrix exponential of A.
-    https://en.wikipedia.org/wiki/Matrix_exponential
-    Note: this is not a general purpose method and is only intended for use within
-    msprime.
+    Only works if the offdiagonals of A are nonnegative
+    and the row sums of A are less than or equal to zero.
+    Melloy & Bennett (1993), https://doi.org/10.1016/0377-0427(93)90036-B
```

Comment:
@petrelharp, I wasn't sure if you'd looked at this ref, or come up with this indepentently. It helped me understand what you were doing anyhow.

Reply:
Oh, great! That not only does just what we do, but proves that it's optimal in some sense! (I worked it out myself, but it's very natural - we're "computing the transition probability of a continuous-time Markov process by Poissonization"; and the reason I did it that way was that Simon Tavaré told me once that that's the best way to do it.)

Note that the reference wants row sums are zero, when in fact it works if they are less than or equal to zero.